cp1616/src/cp1616_io_controller.cpp

```diff
@@ -0,0 +1,510 @@
+
+    return 0;
+  }
+}
+
+int Cp1616IOController::init()
+{
+  PNIO_UINT32 error_code = PNIO_OK;
+
+  //-------------------------------------------------------------------------
+  //Configure Controller Data
+  //-------------------------------------------------------------------------
+  configureControllerData();
+
+  //-------------------------------------------------------------------------
+  //Open PNIO_Controller
+  //-------------------------------------------------------------------------
+
+  //Connect to CP and obtain handle
+  error_code = PNIO_controller_open(
+              cp_id_,
+              PNIO_CEP_MODE_CTRL,
+              &PNIOCbfForDsReadConf,
+              &PNIOCbfForDsWriteConf,
+              &PNIOCbfForAlarmIndication,
+              &cp_handle_);
+
+  //Check errors
+  if(error_code != PNIO_OK)
+  {
+    ROS_ERROR("Not able to open PNIO_controller: Error: 0x%x", (int)error_code);
+    return (int)error_code;
+  }
+  else
+    ROS_INFO_STREAM("Openning CP1616 in IO_controller mode: done");
+
+  //---------------------------------------------------------------------------
+  //register PNIO_CBE_MODE_IND callback for mode changes confirmation
+  //---------------------------------------------------------------------------
+  error_code = PNIO_register_cbf(
+                cp_handle_,
+                PNIO_CBE_MODE_IND,
+                &PNIOCbfForModeChangeIndication);
+
+  //Check errors
+  if(error_code != PNIO_OK)
+  {
+    ROS_ERROR("Error in PNIO_register_cbf: callbackForModeChangeIndication: 0x%x", (int)error_code);
+    PNIO_close(cp_handle_);
+    return (int)error_code;
+  }
+
+  //---------------------------------------------------------------------------
+  //register the callback PNIO_CBE_DEV_ACT for device activation confirmation
+  //---------------------------------------------------------------------------
+  error_code = PNIO_register_cbf(
+                cp_handle_,
+                PNIO_CBE_DEV_ACT_CONF,
+                &PNIOCbfForDeviceActivation);
+
+  if(error_code != PNIO_OK)
+  {
+    ROS_ERROR_STREAM("Error in PNIO_register_cbf: callbackForDeviceActivation: 0x%x" << (int)error_code);
+    PNIO_close(cp_handle_);
+    return (int)error_code;
+  }
+
+  //---------------------------------------------------------------------------
+  //Change CP mode to Operate
+  //---------------------------------------------------------------------------
+  error_code = changeAndWaitForPnioMode(PNIO_MODE_OPERATE);
+  if(error_code != PNIO_OK)
+  {
+      PNIO_close(cp_handle_);
+      return (int)error_code;
+  }
+
+  //---------------------------------------------------------------------------
+  //Start communication
+  //---------------------------------------------------------------------------
+  unsigned i = 0;
+
+  while(cp_ready_ == 0) //Wait for cp_ready flag to be set by callbackForAlarmIndication
+  {
+    i++;
+    usleep(100000);
```

Comment:
I think the idea was more to give it a meaningful name. Something like `CP_READY_WAIT_TIMEOUT` (or something, I have no real idea yet what this code is doing).